src/lightning/pytorch/utilities/model_helpers.py

```diff
@@ -108,18 +107,22 @@ class _restricted_classmethod_impl(Generic[_T, _P, _R_co]):
     """Drop-in replacement for @classmethod, but raises an exception when the decorated method is called on an instance
     instead of a class type."""
 
-    def __init__(self, method: Callable[Concatenate[_T, _P], _R_co]) -> None:
+    def __init__(self, method: Callable[Concatenate[Type[_T], _P], _R_co]) -> None:
         self.method = method
 
     def __get__(self, instance: Optional[_T], cls: Type[_T]) -> Callable[_P, _R_co]:
-        # Workaround for https://github.com/pytorch/pytorch/issues/67146
-        is_scripting = any(os.path.join("torch", "jit") in frameinfo.filename for frameinfo in inspect.stack())
-        if instance is not None and not is_scripting:
-            raise TypeError(
-                f"The classmethod `{cls.__name__}.{self.method.__name__}` cannot be called on an instance."
-                " Please call it on the class type and make sure the return value is used."
-            )
-        return MethodType(self.method, cls)
+        # The wrapper ensures that the method can be inspected, but not called on an instance
+        def wrapper(*args: Any, **kwargs: Any) -> _R_co:
```

Comment:
The docs build complains, I think we need to use `@functools.wraps(self.method)` here